### ISSUE TITLE
undefined symbol `__assert_fail` when build main module

### ISSUE THREAD

**Issue by g1204** [emscripten-core/emscripten#13633]
> Assertion failed: undefined symbol `__assert_fail`. perhaps a side module was not linked in? if this global was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment

I already added the environment var as told but not working. 2.0.15.
didn't have this error in 2.0.8. strongly suspected it's the same reason as issue #13226.

**Comment by sbc100** [comment]
Do you have a repro case perhaps?   __assert_fail is a JS library function (defined in `src/library.js`)

**Comment by Mintyboi** [comment]
I encountered the same issue.
I believe we get this error when we attempt to call a JS function by it's address, e.g. through a function pointer.

I have attached a small sample case which reproduces this issue.
Build command: `emcc foo.cpp -s MAIN_MODULE=2 -s ASSERTIONS=1 -Oz --js-library jslibrary.js -o index.js`
[reproducer.zip](https://github.com/emscripten-core/emscripten/files/6165137/reproducer.zip)

**Comment by sbc100** [comment]
When I run your example I see `undefined symbol `jsPrintHello`.   These seems different to the __assert_fail issue reported.. at least it seems so.

The problem you have is that you are using `MAIN_MODULE=2` ... the error goes away with `MAIN_MODULE=1` .. and also if I do `-s EXPORTED_FUNCTIONS=[_main,_jsPrintHello]`.   It does seem like maybe a bug in `MAIN_MODULE=2`.    @g1204 are you also using `MAIN_MODULE=2`?

**Comment by Mintyboi** [comment]
@sbc100 yes, I should have mentioned that it is a different error message, but I believe it is the same cause where a JS function is added to the GOT but cannot be found.

Building with `MAIN_MODULE=1` or `-s EXPORTED_FUNCTIONS=[_main,_jsPrintHello]` , I can see these lines of codes:
```
function _jsPrintHello() {
    console.log("Hello, world! From JSLand")
}
Module["_jsPrintHello"] = _jsPrintHello;
_jsPrintHello.sig = "v";
```
Notice that we now add the JS function into the module object (which is where `reportUndefinedSymbols()` is attempting to reference from). Same goes for `__assert_fail`

**Comment by Mintyboi** [comment]
What I did to make it work in my code was to edit `resolveGlobalSymbol()` to include looking for the JS function in the global namespace.
It's ugly, and inefficient in node env, but it works as a patch. @sbc100 do you know of a more elegant way?
```
function resolveGlobalSymbol(symName, direct) {
    var sym;
    if (direct) {
        sym = Module["asm"]["orig$" + symName]
    }
    if (!sym) {
        sym = Module["asm"][symName]
    }
    if (!sym && direct) {
        sym = Module["_orig$" + symName]
    }
    if (!sym) {
        sym = Module[asmjsMangle(symName)]
    }
    if (!sym) {
        if (ENVIRONMENT_IS_NODE) eval(`sym = typeof(${asmjsMangle(symName)}) !== "undefined" ? ${asmjsMangle(symName)} : undefined;`);
        else sym = globalThis[asmjsMangle(symName)]
    }
    if (!sym && symName.indexOf("invoke_") == 0) {
        sym = createInvokeFunction(symName.split("_")[1])
    }
    return sym
}
```

**Comment by sbc100** [comment]
Lets wait to see if @g1204 is using `MAIN_MODULE=2` .. if they are then we have found the issue.. if not its a different problem.

As to how to fix it, I'd need to take closer look.. I'm not sure we want the global namespace taking part in the linking process, but maybe its the right direction..

**Comment by Mintyboi** [comment]
@sbc100 I have created another issue for my issue in https://github.com/emscripten-core/emscripten/issues/13786 so that we can focus our discussions on what I'm seeing.
